### PR TITLE
PLAT-3280: Return placeholder images instead of 404 errors

### PR DESCRIPTION
**Pull request by RichardLeeWattpad**
@jharlap

### CODE REVIEW

File: src/config.js [hunk 1]
```diff
@@ -68,6 +68,7 @@ exports.transformRedirectOnError = process.env.KAMU_TRANS_REDIRECT_ONERROR ? pro
 
 // default safe headers to be used by default on responses
 exports.defaultHeaders  = {
+  "Content-Type": "text",
```
Comment (jharlap) [inline src/config.js]: why?

File: src/utils/index.js [hunk 2]
```diff
@@ -36,8 +38,15 @@ var writeErrorHead = function( res, statusCode ) {
  */
 var fourOhFour = function( res, msg, url ) {
   log.warn( msg + ': ' + ( ( url != null ? url.format() : void 0 ) || 'unknown' ) );
-  writeErrorHead( res, 404 );
-  return finish( res, 'Not Found' );
+
+  try {
+    var data = fs.readFileSync('./placeholders/404.jpg');
```
Comment (jharlap) [inline src/utils/index.js]: blocking file i/o will prevent kamu from responding to other requests in parallel

Comment (jessicawoo) [inline src/utils/index.js]: Is there a reason we aren't using readFile and a callback, VS readFileSync?

File: src/utils/index.js [hunk 3]
```diff
@@ -15,8 +16,9 @@ var config      = require( '../config' ),
  * @param     res           object        response object
  * @param     statusCode    integer       numeric response code to write
  */
-var writeErrorHead = function( res, statusCode ) {
+var writeErrorHead = function( res, statusCode, contentType ) {
```
Comment (jessicawoo) [inline src/utils/index.js]: This method name confuses me, it doesn't actually write an 'Error'.

File: src/utils/index.js [hunk 4]
```diff
@@ -36,8 +38,15 @@ var writeErrorHead = function( res, statusCode ) {
...
   log.warn( msg + ': ' + ( ( url != null ? url.format() : void 0 ) || 'unknown' ) );
-  writeErrorHead( res, 404 );
-  return finish( res, 'Not Found' );
+
+  try {
+    var data = fs.readFileSync('./placeholders/404.jpg');
+    writeErrorHead( res, 200, 'image/jpeg' );
+    return finish( res, data );
+  } catch (e) {
+    writeErrorHead( res, 200 );
```
Comment (jessicawoo) [inline src/utils/index.js]: Do we really want to return 200 OK here?

File: test/integration/index.js [hunk 5]
```diff
@@ -240,9 +240,10 @@ describe( 'app integration ->', function() {
         reqOptions.path = bkpPath;
       } );
 
-      it( 'should return a 404', function( done ) {
+      it( 'should return a placeholder image', function( done ) {
         req = http.request( reqOptions, function( res ) {
-          expect( res.statusCode ).to.be.equal( 404 );
+          expect( res.statusCode ).to.be.equal( 200 );
+          expect( res.headers['content-type'] ).to.be.equal( 'image/jpeg' );
```
Comment (jessicawoo) [inline test/integration/index.js]: You can technically test that the 404 image is actually there, no?

File: src/utils/index.js (hunk 3 above)
Comment (RichardLeeWattpad) [inline src/utils/index.js]: It writes the header of an error.

**Comment by RichardLeeWattpad** [comment]
@jharlap @minosabaki Addressed comments.

File: src/utils/index.js [hunk 6]
```diff
@@ -15,8 +16,8 @@ var config      = require( '../config' ),
  * @param     res           object        response object
  * @param     statusCode    integer       numeric response code to write
  */
-var writeErrorHead = function( res, statusCode ) {
-  res.writeHead( statusCode, {
+var writeErrorHead = function( res, statusCode, contentType ) {
+  var headers = {
```
Comment (jessicawoo) [inline src/utils/index.js:23]: Still haven't renamed the function.

File: src/utils/index.js [hunk 7]
```diff
@@ -24,20 +25,38 @@ var writeErrorHead = function( res, statusCode ) {
...
 /*
- * respond with a 404 response
+ * for backwards compatibility with legacy clients, respond with a placeholder image instead of a normal 404 response
  *
  * @param   object    res     http response object
  * @param   string    msg     404 message string
  * @param   object    url     url object
  */
-var fourOhFour = function( res, msg, url ) {
+var fourOhFour = function( res, msg, url, callback ) {
```
Comment (jessicawoo) [inline src/utils/index.js]: By callback, I didn't mean the function takes a callback. There's no reason to. 

When you execute readFile, the anonymous function you attach as the second parameter IS the callback.

**Comment by RichardLeeWattpad** [comment]
@minosabaki Ready for review.

File: src/utils/index.js [hunk 8]
```diff
@@ -9,14 +10,17 @@ var config      = require( '../config' ),
     log         = require( '../log' ),
     connStatus  = require( '../connection' );
 
+// Placeholder image to show when 404
+var placeholderImage = fs.readFileSync( './placeholders/404.jpg' );
```
Comment (jessicawoo) [inline src/utils/index.js]: So this filepath, while I suppose it should work, it heavily depends on the fact that the nodejs server execution command is ran inside of the base project folder. A more dependable way of doing this is to use __dirname, which gives you the current directory that this JS file resides in, and relatively referencing this file.

File: src/utils/index.js [hunk 9]
```diff
@@ -9,14 +10,17 @@ var config      = require( '../config' ),
...
  * write the error response headers
  *
  * @param     res           object        response object
  * @param     statusCode    integer       numeric response code to write
  */
-var writeErrorHead = function( res, statusCode ) {
-  res.writeHead( statusCode, {
+var writeHeadersWithStatusAndContentType = function( res, statusCode, contentType ) {
+  var headers = {
     'expires': '0',
```
Comment (jessicawoo) [inline src/utils/index.js:24]: This should be capitalized to be consistent with the rest that we have listed below, and it should not be a zero value; it should be set to some value in the past. Remember to format the date correctly.

File: test/integration/index.js [hunk 10]
```diff
@@ -273,10 +282,16 @@ describe( 'app integration ->', function() {
...
         } );
 
-        it( 'should return a 404', function( done ) {
+        it( 'should return a placeholder image', function( done ) {
           req = http.request( reqOptions, function( res ) {
-            expect( res.statusCode ).to.be.equal( 404 );
-            done();
+            expect( res.statusCode ).to.be.equal( 200 );
+            expect( res.headers[ 'content-type' ] ).to.be.equal( 'image/jpeg' );
+
```
Comment (jessicawoo) [inline test/integration/index.js]: Extra space?

File: test/integration/index.js [hunk 11]
```diff
@@ -240,10 +243,16 @@ describe( 'app integration ->', function() {
...
+      it( 'should return a placeholder image', function( done ) {
         req = http.request( reqOptions, function( res ) {
-          expect( res.statusCode ).to.be.equal( 404 );
-          done();
+          expect( res.statusCode ).to.be.equal( 200 );
+          expect( res.headers[ 'content-type' ] ).to.be.equal( 'image/jpeg' );
+
+          res.on( 'data', function( chunk ) {
+            expect( JSON.stringify( chunk ) ).to.be.equal( JSON.stringify( placeholderImage ) );
+
```
Comment (jessicawoo) [inline test/integration/index.js]: Space?

File: test/integration/index.js [hunk 12]
```diff
@@ -240,10 +243,16 @@ describe( 'app integration ->', function() {
...
       } );
 
-      it( 'should return a 404', function( done ) {
+      it( 'should return a placeholder image', function( done ) {
         req = http.request( reqOptions, function( res ) {
-          expect( res.statusCode ).to.be.equal( 404 );
-          done();
+          expect( res.statusCode ).to.be.equal( 200 );
+          expect( res.headers[ 'content-type' ] ).to.be.equal( 'image/jpeg' );
+
```
Comment (jessicawoo) [inline test/integration/index.js]: Space?

File: test/integration/index.js [hunk 13]
```diff
@@ -273,10 +282,16 @@ describe( 'app integration ->', function() {
...
+        it( 'should return a placeholder image', function( done ) {
           req = http.request( reqOptions, function( res ) {
-            expect( res.statusCode ).to.be.equal( 404 );
-            done();
+            expect( res.statusCode ).to.be.equal( 200 );
+            expect( res.headers[ 'content-type' ] ).to.be.equal( 'image/jpeg' );
+
+            res.on( 'data', function( chunk ) {
+              expect( JSON.stringify( chunk ) ).to.be.equal( JSON.stringify( placeholderImage ) );
+
```
Comment (jessicawoo) [inline test/integration/index.js]: Space?

File: test/integration/index.js [hunk 14]
```diff
@@ -297,10 +312,15 @@ describe( 'app integration ->', function() {
         } );
 
 
-        it( 'should return a 404', function( done ) {
+        it( 'should return a placeholder image', function( done ) {
           req = http.request( reqOptions, function( res ) {
-            expect( res.statusCode ).to.be.equal( 404 );
-            done();
+            expect( res.headers[ 'content-type' ] ).to.be.equal( 'image/jpeg' );
+
```
Comment (jessicawoo) [inline test/integration/index.js]: Space?

File: test/integration/index.js [hunk 15]
```diff
@@ -297,10 +312,15 @@ describe( 'app integration ->', function() {
...
+            res.on( 'data', function( chunk ) {
+              expect( JSON.stringify( chunk ) ).to.be.equal( JSON.stringify( placeholderImage ) );
+
```
Comment (jessicawoo) [inline test/integration/index.js]: Alright you get it now =)

File: test/specs/src/utils/index.spec.js [hunk 16]
```diff
@@ -68,7 +68,7 @@ describe( 'utils index', function() {
     } );
 
     after( function() {
-      revertWriteErrorHead();
+      revertwriteHeadersWithStatusAndContentType();
```
Comment (jessicawoo) [inline test/specs/src/utils/index.spec.js]: Capitalize the W.

**Comment by jessicawoo** [comment]
@RichardLeeWattpad review done

File: test/integration/index.js (hunk 15 above)
Comment (jharlap) [inline test/integration/index.js]: if the content type is image/jpeg, then the data chunk will be a block out of a binary image file, so what's the purpose in stringifying it before doing the comparison? (that's actually a question, not just a veiled way to say "don't stringify", although if there isn't a good reason, then don't :wink:)

File: src/utils/index.js (hunk 9 above)
Comment (RichardLeeWattpad) [inline src/utils/index.js:24]: This is Marcos' code, I'm really not sure how much we should be changing on him.

File: test/integration/index.js (hunk 10 above)
Comment (RichardLeeWattpad) [inline test/integration/index.js]: I wanted to put a line break in between expect()'s and done()'s.

File: test/integration/index.js (hunk 15 above)
Comment (RichardLeeWattpad) [inline test/integration/index.js]: The comparison didn't work when comparing the binaries.

**Comment by RichardLeeWattpad** [comment]
@minosabaki @jharlap Addressed comments except for the expires one.

File: test/integration/index.js [hunk 17]
```diff
@@ -4,9 +4,12 @@ require( '../common' );
 
 var http      = require( 'http' ),
     Url       = require( 'url' ),
+    fs        = require( 'fs' ),
     proxyUrl  = require( '../../demo/utils' ).proxyUrl,
     server;
 
+var placeholderImage = fs.readFileSync( './placeholders/404.jpg' );
```
Comment (jessicawoo) [inline test/integration/index.js]: Same here with the relative pathing

File: test/integration/index.js [hunk 18]
```diff
@@ -240,10 +243,14 @@ describe( 'app integration ->', function() {
...
 
-      it( 'should return a 404', function( done ) {
+      it( 'should return a placeholder image', function( done ) {
         req = http.request( reqOptions, function( res ) {
-          expect( res.statusCode ).to.be.equal( 404 );
-          done();
+          expect( res.statusCode ).to.be.equal( 200 );
+          expect( res.headers[ 'content-type' ] ).to.be.equal( 'image/jpeg' );
+          res.on( 'data', function( chunk ) {
+            expect( chunk.equals(placeholderImage) ).to.be.equal( true );
```
Comment (jessicawoo) [inline test/integration/index.js]: wait so " expect( chunk ).to.eql( placeholderImage )" doesn't return true?

**Comment by jessicawoo** [comment]
R+, squash, rebase and :shipit:

**Pull request merged**